Linux VRF Route Lookup Fallback Control HLD

sudharr86 opened the pull request:
This document provides the high-level design for Linux VRF Route Lookup Fallback Control. By default, `vrfmgrd` blocks legacy Linux RPDB fall-through in applicable non-default VRFs, aligning the kernel route-miss result with ASICs that terminate a VRF miss. An explicit global compatibility setting can temporarily restore legacy kernel fallback. The feature changes Linux kernel routing only; SAI and ASIC programming remain unchanged.

sudharr86 committed:
No Default VRF Route Fallback HLD

Signed-off-by: Sudharsan Rajagopalan <sudharr@cisco.com>

sudharr86 committed:
Update no-default-vrf-route-fallback-hld.md

Signed-off-by: Sudharsan Rajagopalan <sudharr@cisco.com>

mssonicbld commented:
/azp run

sudharr86 committed:
Update no-default-vrf-route-fallback-hld.md

Signed-off-by: Sudharsan Rajagopalan <sudharr@cisco.com>

mssonicbld commented:
/azp run

rck-innovium commented on doc/vrf/no-default-vrf-route-fallback-hld.md:
```diff
@@ -0,0 +1,411 @@
+
+**No SAI API changes are required for this feature.**
+
+The unreachable routes are maintained only in the Linux kernel and are explicitly not programmed into the ASIC hardware. This design choice ensures:
```
This may not be true. Consider the below config:
```
"VRF_ROUTE_FALLBACK_GLOBAL|ipv4": {
    "status": "enabled"
},
"VRF|VrfRed": {
    "fallback": "enabled",
},
"VRF|VrfGreen": {
    "fallback": "disabled",
}
```
In this case, in the ASIC route tables, **VrfGreen|0.0.0.0/0** route needs to be installed to avoid fallback VRF lookup for packets in VrfGreen.

rck-innovium replied:
Can you please explain what is the relation of the proposed feature with SAI_SWITCH_ATTR_DEFAULT_VIRTUAL_ROUTER_ID

sudharr86 committed:
Update no-default-vrf-route-fallback-hld.md

Default unreachable route, global fallback knob added

Signed-off-by: Sudharsan Rajagopalan <sudharr@cisco.com>

sudharr86 committed:
Update no-default-vrf-route-fallback-hld.md

Signed-off-by: Sudharsan Rajagopalan <sudharr@cisco.com>

mssonicbld commented:
/azp run

ashutosh-agrawal commented on doc/vrf/no-default-vrf-route-fallback-hld.md:
```diff
@@ -0,0 +1,334 @@
+
+### 7.2. Global Fallback Knob
+
+- New global configuration (in CONFIG_DB and CLI): `KERNEL_VRF_FALLBACK`
```
IIRC, in the HLD review meeting, we agreed to implement only the CONFIG_DB mechanism and not the CLI.

sudharr86 replied:
Addressed in the current HLD. Sections 2, 5.3 EX-3, 7.3, and 9 explicitly define CONFIG_DB-only management and state that no feature-specific CLI, show command, or management API is introduced.

ashutosh-agrawal commented on doc/vrf/no-default-vrf-route-fallback-hld.md:
```diff
@@ -0,0 +1,334 @@
+### 9.3. Backward Compatibility
+
+- Default (knob unset): fallback disabled, matches ASIC.
+- Knob enabled: fallback restored (legacy only).
```
Please change to "fallback restored in the kernel forwarding path only"

sudharr86 replied:
agreed.

sudharr86 replied:
Addressed in the current HLD. Sections 2 and 7.1 now state that status=enabled restores legacy RPDB fall-through only in the Linux kernel route-lookup path; APP_DB, ASIC_DB, SAI, and hardware behavior remain unchanged.

gord1306 commented:
I've submitted a similar PR before. Here's just a link for reference: https://github.com/sonic-net/sonic-swss/pull/2943

sudharr86 commented on doc/vrf/no-default-vrf-route-fallback-hld.md:
```diff
@@ -0,0 +1,411 @@
+
+## 2. Scope
+
+This high-level design document covers the implementation of a configurable mechanism to disable VRF (Virtual Routing and Forwarding) route fallback behavior in SONiC. The scope includes:
```
agreed.

sudharr86 replied:
Addressed in the current HLD. Section 2 limits the change to Linux VRF route lookup for host-local and software-forwarded traffic, and Sections 4 and 8 explicitly state that ASIC, SAI, and hardware forwarding behavior are unchanged.

sudharr86 committed:
Address VRF fallback HLD review comments

Signed-off-by: Sudharsan Rajagopalan <sudharr@cisco.com>

sudharr86 committed:
Clarify VRF fallback defaults and migration

Signed-off-by: Sudharsan Rajagopalan <sudharr@cisco.com>

sudharr86 committed:
Clarify VRF fallback namespace behavior

Document applicable VRFs, namespace-local multi-ASIC persistence, and safe rollback ordering discovered during implementation.

Signed-off-by: Sudharsan Rajagopalan <sudharr@cisco.com>

sudharr86 commented on doc/vrf/no-default-vrf-route-fallback-hld.md:
```diff
@@ -0,0 +1,334 @@
+In prior SONiC releases, Linux kernel allowed VRF lookup fallback if a route was missing in a non-default VRF, which could cause divergence from ASIC behavior and security/operational risks.
+
+**This design ensures:**
+- By default, fallback is disabled both in kernel and ASIC by installing an unreachable default route in every new non-default VRF.
```
Addressed in the current HLD. Section 4 and Section 7.4 state that fpmsyncd filters the RTN_UNREACHABLE notifications before APP_DB mutation, so the managed sentinels do not enter APP_DB, ASIC_DB, SAI, or the hardware FIB.

sudharr86 committed:
Rename VRF fallback control feature

Use Linux VRF Route Lookup Fallback Control for the HLD title and file name.

Signed-off-by: Sudharsan Rajagopalan <sudharr@cisco.com>
